### ISSUE TITLE
Schema v2 tracker

### ISSUE THREAD

**Issue by dbecvarik** [cekit/cekit#206]
This issue is here for tracking schema changes.

# 1. repository for yum packages
``` yaml
packages:
      repositories:
           - jboss-os
```
will be changed to defining repositories as a full objects as follows
``` yaml
packages:
      repositories:
            - name: rhel-7-server-ose-3.7-rpms
```
where repository name must be a real name of puddle content set which will be created by ODCS.

**Comment by rcernich** [comment]
How would folks in the community use this?  How would this work for building against centos, fedora, etc.?

**Comment by dbecvarik** [comment]
In ideal way, centos/fedora images will be upstream for us, so we will ad repositories in -rhel specific modules/overrides. Which is preferred option. The biggest difference is that we are not disabling installed repositories anymore, so most of the centos images should just work. This started to cause issues with OSBS, because OSBS is injecting repositories which are breaking our images if we do not specify proper one via content sets and ODCS integration.

For short term goal, you will be able to disable usage of ODCS and then these repositories will be ignored and repositories from image will be used.  If you need additional repository for community you need to install it via rpm as is a standard way for EPEL and others (Issue is, you need GPG key alongside with the repo to use it).

Also there is different handling for local docker build (directly calling ODCS and consuming repo files) and OSBS which is using container.yaml file to construct content sets for us.

everything will be configurable in ~/.cekit/config file in repository section like

```
[repository]
builder-docker = odcs-puddle # use ODCS to construct puddle and get repo files
builder-docker = ignore # do not handle files in community
builder-osbs = osbs-contentset # use container.yaml to tell OSBS which packages in content set are needed

```

This syntax is still not final. It's pretty hard to find way which is compatible with our stuff and provide easy migration to community enabled solution. Especially in the phase, where our upstream unfriendly approach of downloading .repo files was moved from image descriptors to modules ones.
The approach is, that you will have repositories defined in image descriptors, and in your Cekit config files you will configure some kind of a "profile" which will tell Cekit what to do with repositories as we need to support a lot of incompatible ways of repository handling.

**Comment by rcernich** [comment]
Beyond community images, there is the larger community of people.  How would people outside Red Hat use this solution?  This is one of the problems I have with the repo solution: there's not really anything that says what jboss-os means or what should be configured in order to make the build succeed.

**Comment by dbecvarik** [comment]
Its little bit better with new solution, where repository name should be name of the existing RHEL repository. But this is still causing troubles and we need to remove any RHEL specific repositories from our code and handle this in RHEL specific modules or overrides. So we need to build upstream images first and then convert them to RHEL ones, as RHEL ones requires special approach.  
We must  change our mind and think about all stuff we introduce to overcome OSBS/RHEL subscription issues to be **deviation** and implement this as layering on community images. Doing it opposite as now is causing a lot of troubles and headaches.

**Comment by dbecvarik** [comment]
The important point is that lot of our images doesn't require additional repository for centos/fedora images, but our internal environment requires us to inject repos like jboss-os which is just base repo in centos world, which is available inside the image.

**Comment by rcernich** [comment]
Sorry, the point I was getting at was, how would somebody outside Red Hat build the image?  We can move things around to build community versions of images (e.g. WildFly on CentOS), but we also need to allow folks to build these on their own.

I also think including OSBS details in here convolutes the tool, as this should be specific to building the image and particular details related to downstream build management should be in a separate tool, IMHO, but this is off-topic.

**Comment by goldmann** [comment]
We should move this discussion out of this ticket. @dbecvarik what is the ticket to track ODCS work?

@rcernich Proposed change is not to base only on ODCS, but extend the list of possible ways of providing the RPM content by adding ODCS. This is a major change and it requires to make the image definition not related to the delivery mechanism. This **is** a tricky thing to solve properly, that's why @dbecvarik is spending so much time on it. We believe that this is make the tool better.

> Beyond community images, there is the larger community of people. How would people outside Red Hat use this solution? This is one of the problems I have with the repo solution: there's not really anything that says what jboss-os means or what should be configured in order to make the build succeed.

And this is where the proposed solution is better than what we have now. The difference is that the person who wants to rebuild the image exactly knows which repository needs to be provided so that the build succeeds. It can be done in many different ways, this is where Cekit documentation will help. You can enable these on the host where you build or you can point it to **some** internal repo. How customer wants to do this depends on what they have available. We can only guess. I suggest to wait for @dbecvarik to put a document describing the proposed changes.

> I also think including OSBS details in here convolutes the tool, as this should be specific to building the image and particular details related to downstream build management should be in a separate tool, IMHO, but this is off-topic.

I don't agree. OSBS is just a builder, an upstream one that could be hosted by anyone (https://osbs.readthedocs.io/en/latest/). Same applies to ODCS (https://pagure.io/odcs/). These are the tools that are used internally, but there is nothing that could prevent anyone from running these. These are not specific downstream tools. If someone is interested to replicate our build system, it can be done.

**Comment by rcernich** [comment]
> > I also think including OSBS details in here convolutes the tool, as this should be specific to building the image and particular details related to downstream build management should be in a separate tool, IMHO, but this is off-topic.

> I don't agree. OSBS is just a builder, an upstream one that could be hosted by anyone (https://osbs.readthedocs.io/en/latest/). Same applies to ODCS (https://pagure.io/odcs/). These are the tools that are used internally, but there is nothing that could prevent anyone from running these. These are not specific downstream tools. If someone is interested to replicate our build system, it can be done.

I think having a separate tool, that manages the OSBS build process is a better solution.  This tool would invoke cekit to perform the build, while managing all of the other things required for OSBS.  This would allow the OSBS configuration to be defined externally to the image build details.

**Comment by dbecvarik** [comment]
New repositories handling is implemented in #202
